cmd/tailscale/cli/up.go

```diff
@@ -1102,3 +1111,96 @@ func anyPeerAdvertisingRoutes(st *ipnstate.Status) bool {
 	}
 	return false
 }
+
+// resolveAuthKey either returns v unchanged (in the common case)
+// or, if it starts with "oauth:" parses it as one of:
+//
+//	oauth2:CLIENT_ID:CLIENT_SECRET?ephemeral=BOOL&tags=foo,bar&preauthorized=BOOL
+//	oauth2:CLIENT_ID:CLIENT_SECRET:BASE_URL?...
+//
+// and does the OAuth2 dance to get and return an authkey.
+//
+// If the BASE_URL argument is not provided, it defaults to https://api.tailscale.com.
+func resolveAuthKey(ctx context.Context, v string) (string, error) {
+	suff, ok := strings.CutPrefix(v, "oauth:")
+	if !ok {
+		return v, nil
+	}
+	if !envknob.Bool("TS_EXPERIMENT_OAUTH_AUTHKEY") {
+		return "", errors.New("oauth authkeys are in experimental status")
+	}
+
+	pos, named, _ := strings.Cut(suff, "?")
+	attrs, err := url.ParseQuery(named)
+	if err != nil {
+		return "", err
+	}
+	for k := range attrs {
+		switch k {
+		case "ephemeral", "preauthorized", "tags":
+		default:
+			return "", fmt.Errorf("unknown attribute %q", k)
+		}
+	}
+	getBool := func(name string, def bool) (bool, error) {
+		v := attrs.Get(name)
+		if v == "" {
+			return def, nil
+		}
+		ret, err := strconv.ParseBool(v)
+		if err != nil {
+			return false, fmt.Errorf("invalid attribute boolean attribute %s value %q", name, v)
+		}
+		return ret, nil
+	}
+	ephemeral, err := getBool("ephemeral", false)
```

Comment:
maybe this should default to true, @DentonGentry?

Reply:
That seems like a safer default if one is automatically generating authkeys. Otherwise people spin up a couple thousand nodes and only figure it out later, then ask us for help in deleting them.